File: web/src/main/webapp/js/com/openclinica/renderer/ItemDefRenderer.js
function ItemDefRenderer(json, itemDetails, mandatory, formOID, repeatRowNumber,studyEventOID,studyEventRepeatKey) {
  this.json = json;
  this.itemDetails = itemDetails;
  this.mandatory = mandatory;
  this.name = this.itemDetails["OpenClinica:LeftItemText"];
  this.rightItemText = this.itemDetails["OpenClinica:RightItemText"];
  this.dataType = json["@DataType"];
  this.responseType = this.itemDetails["OpenClinica:ItemResponse"]["@ResponseType"];
  this.responseLayout = this.itemDetails["OpenClinica:ItemResponse"]["@ResponseLayout"];
  this.isInline = this.responseLayout == "Horizontal" ? "inline" : "";
  this.OID = json["@OID"];
  this.itemName = json["@Name"];
  this.description=json["@Comment"]; 
  this.itemNumber = json["Question"] && json["Question"]["@OpenClinica:QuestionNumber"] ? json["Question"]["@OpenClinica:QuestionNumber"] : "";
  this.unitLabel = json["MeasurementUnitRef"] ? "("+app_basicDefinitions[json["MeasurementUnitRef"]["@MeasurementUnitOID"]]+")" : "";
  this.codeListOID = json["CodeListRef"] ? json["CodeListRef"]["@CodeListOID"] : "";
  this.multiSelectListOID = json["OpenClinica:MultiSelectListRef"] ? json["OpenClinica:MultiSelectListRef"]["@MultiSelectListID"] : "";
  debug("In ItemDefRenderer: " + this.OID + "  multiSelectListOID: " + this.multiSelectListOID, util_logDebug);
  this.columns = this.itemDetails["OpenClinica:Layout"] ? this.itemDetails["OpenClinica:Layout"]["@Columns"] : undefined;
  this.itemValue = undefined;
  this.file = undefined;
  this.fileDownloadLink=undefined;
  this.itemTemp= undefined;
  this.audits = undefined;
  this.dns = undefined;
  this.itemNameLink = undefined;
  this.studyEventOID = studyEventOID;
  this.studyEventRepeatKey = studyEventRepeatKey;
  this.dash="-";
  this.key=this.studyEventOID+this.dash+this.studyEventRepeatKey+this.dash+this.OID;
  this.keyAndRepeat=this.key+this.dash+repeatRowNumber;
  
  
  
  if (this.studyEventOID!=undefined && app_itemValuesMap[this.keyAndRepeat]) { 
    this.itemValue = app_itemValuesMap[this.keyAndRepeat]; 
    
    }
  
  if(this.studyEventOID!=undefined && app_displayAudits=='y' && app_audits[this.key])
	    this.audits = app_audits[this.key];
	    if(this.studyEventOID!=undefined && app_displayDNs=='y' && app_dns[this.key])
	    this.dns = app_dns[this.key];
  
  
  if (this.responseType!=undefined) {
    if (this.responseType=='file' && this.itemValue) {
      if (this.itemValue.indexOf("/")==0) {
        this.fileDownloadLink=app_contextPath+"/DownloadAttachedFile?fileName="+this.itemValue;
        this.file = this.itemValue.substring(this.itemValue.lastIndexOf('/')+1);
        this.fileDownloadLink = this.fileDownloadLink.replace(this.file,  encodeURIComponent(this.file));
      }
      else {
        this.fileDownloadLink=app_contextPath+"/DownloadAttachedFile?fileName="+this.itemValue.replace(/\\/g,"//");
        this.file = this.itemValue.substring(this.itemValue.lastIndexOf('\\')+1);
        this.fileDownloadLink = this.fileDownloadLink.replace(this.file,  encodeURIComponent(this.file));
      }
    }
 }  
  
  this.renderPrintableItem = function(isRepeating) { 
    var template = "print_item_def";
    if (isRepeating == true) {
      template = this.responseLayout == "Horizontal" ? "print_repeating_item_horiz" : "print_repeating_item";
    }
    
		  if (app_thisFormData == undefined || app_thisStudyEvent==undefined ||app_thisSubjectsData==undefined) {
	var itemNameLink = this.itemName
	}else{
	var itemNameLink = app_thisSubjectsData["@SubjectKey"]+"/"+app_thisStudyEvent["@StudyEventOID"]+"["+app_thisStudyEvent["@StudyEventRepeatKey"]+"]/"+ app_thisFormData["@FormOID"]+"/"+this.OID;
	}

    if(app_displayAudits=='y' || app_displayDNs=='y' ){
           var s = RenderUtil.render(RenderUtil.get(template), 
           {OID:this.OID, formOID:formOID, itemNumber:this.itemNumber, name:this.name, rightItemText:this.rightItemText, responseType:this.responseType, 
            unitLabel:this.unitLabel, optionNames: app_codeLists[this.codeListOID], multiSelectOptionNames: app_multiSelectLists[this.multiSelectListOID], 
            columns:this.columns, responseLayout:this.responseLayout, isInline: this.isInline, mandatory: this.mandatory,
            itemNameLink:itemNameLink, itemName:this.itemName, itemValue:this.itemValue,file:this.file,fileDownloadLink:this.fileDownloadLink});
          }else{
        var s = RenderUtil.render(RenderUtil.get(template), 
           {OID:this.OID, formOID:formOID, itemNumber:this.itemNumber, name:this.name, rightItemText:this.rightItemText, responseType:this.responseType, 
            unitLabel:this.unitLabel, optionNames: app_codeLists[this.codeListOID], multiSelectOptionNames: app_multiSelectLists[this.multiSelectListOID], 
            columns:this.columns, responseLayout:this.responseLayout, isInline: this.isInline, mandatory: this.mandatory,
            itemValue:this.itemValue,file:this.file,fileDownloadLink:this.fileDownloadLink});
    }	
    return s[0].outerHTML;
  }
  
  
  this.renderItemFormMetadata = function(groupLabel){
    var template = "print_item_metadata_info";
    var subjectOID = app_thisSubjectsData["@SubjectKey"];
    var responseOptions =  app_codeLists[this.codeListOID]?app_codeLists[this.codeListOID]:app_multiSelectLists[this.multiSelectListOID];

    if (app_thisFormData == undefined || app_thisStudyEvent==undefined ||app_thisSubjectsData==undefined) {
	var itemNameLink =this.itemName
	}else{
	var itemNameLink = app_thisSubjectsData["@SubjectKey"]+"/"+app_thisStudyEvent["@StudyEventOID"]+"["+app_thisStudyEvent["@StudyEventRepeatKey"]+"]/"+ app_thisFormData["@FormOID"]+"/"+this.OID;
	}

    var s = RenderUtil.render(RenderUtil.get(template), 
           {itemNameLink:itemNameLink, itemName:this.itemName,leftItemText:this.name,units:this.unitLabel,responseOptions:responseOptions,dataType:this.dataType ,description:this.description,groupLabel:groupLabel});
        return s[0].outerHTML;
        
  }
  this.renderAuditLogs = function(auditLogs,repeatRowNumber,repeating){
    var template="print_audits";
    var s = RenderUtil.render(RenderUtil.get(template),{auditLogs:auditLogs,value:this.itemName,repeatRow:repeatRowNumber,repeating:repeating});
    return s[0].outerHTML;
  }
   
this.renderDiscrepancyNotes = function(discrepancyNotes,repeatRowNumber,repeating){
    var template="print_dns";
    var s = RenderUtil.render(RenderUtil.get(template),{discrepancyNotes:discrepancyNotes,value:this.itemName,repeatRow:repeatRowNumber, repeating:repeating});
    return s[0].outerHTML;
  }

   
}

File: web/src/main/java/org/akaza/openclinica/renderer/ItemDefRenderer.java
package org.akaza.openclinica.renderer;

import java.io.StringWriter;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import freemarker.template.Configuration;
import freemarker.template.Template;
import net.sf.json.JSON;
import net.sf.json.JSONArray;
import net.sf.json.JSONObject;

public class ItemDefRenderer extends JSONRenderer{
  public JSONObject itemDef;
  private JSONObject itemDetails;
  private boolean mandatory;
  private String template;
  private String name;
  private String rightItemText;
  private String dataType;
  private String responseType;
  private String responseLayout;
  private String isInline;
  private String OID;
  private String itemName;
  private String itemNumber;
  private String unitLabel;
  private String codeListOID;
  private String multiSelectListOID;
  private int columns;
  private  Map<String, ItemGroup> appItemGroupDefs = new TreeMap<String,ItemGroup>();
  private  Map<String, String> appItemGroupMap = new TreeMap<String,String>();
  private  Map<String, String> appBasicDefinitions = new TreeMap<String,String>();
  private  Map<String, List> appCodeLists = new TreeMap<String,List>();
  private  Map<String, List> appMultiSelectLists = new TreeMap<String,List>();
  
  public ItemDefRenderer(JSONObject itemDef, Configuration cfg, Map templateVars, 
                         Map appItemGroupDefs, Map appItemGroupMap,
                         Map appBasicDefinitions, Map appCodeLists, Map appMultiSelectLists) {
    super(itemDef, cfg, templateVars);
    this.itemDef = itemDef;
    this.appItemGroupDefs = appItemGroupDefs;
    this.appItemGroupMap = appItemGroupMap;
    this.appBasicDefinitions = appBasicDefinitions;
    this.appCodeLists = appCodeLists;
    this.appMultiSelectLists = appMultiSelectLists;
  }
  
  
  public String parse(JSONObject itemDef)  {
    itemDetails = itemDef.getJSONObject("OpenClinica:ItemDetails");
    JSONArray presentInForm = StudyDataLoader.ensureArray(itemDef.getJSONObject("OpenClinica:ItemDetails"), "OpenClinica:ItemPresentInForm");
    OID = itemDef.optString("@OID");
    dataType = itemDef.optString("@DataType");
    name = itemDetails.optString("OpenClinica:LeftItemText");
    itemName = itemDetails.optString("@Name");
    responseType = itemDetails.getJSONObject("OpenClinica:ItemPresentInForm").getJSONObject("OpenClinica:ItemResponse").optString("@ResponseType");
    responseLayout = itemDetails.getJSONObject("OpenClinica:ItemPresentInForm").getJSONObject("OpenClinica:ItemResponse").optString("@ResponseLayout");
    isInline = responseLayout == "Horizontal" ? "inline" : "";
    rightItemText = itemDetails.optString("OpenClinica:RightItemText");
    itemNumber = itemDef.getJSONObject("Question").has("@OpenClinica:QuestionNumber") ? itemDef.getJSONObject("Question").getString("@OpenClinica:QuestionNumber") + "." : "";
    unitLabel = itemDef.has("MeasurementUnitRef") ? appBasicDefinitions.get(itemDef.getJSONObject("MeasurementUnitRef").getString("@MeasurementUnitOID")) : ""; 
    codeListOID = itemDef.has("CodeListRef") ? itemDef.getJSONObject("CodeListRef").getString("@CodeListOID") : "";
    multiSelectListOID = itemDef.has("MultiSelectListRef") ? itemDef.getJSONObject("MultiSelectListRef").getString("@MultiSelectListOID") : "";
    //JSONObject question = ((JSONArray)presentInForm).getJSONObject(1);
    //columns = question.has("OpenClinica:Layout") ? question.getJSONObject("OpenClinica:Layout").getInt("@Columns") : 2; 
    
   return "";
  }

  public String renderPrintableItem(boolean isRepeating)  {
    parse(this.itemDef);
    String  template = "item_def.ftl";
    
    if (isRepeating == true) {
      template = responseLayout == "Horizontal" ? "repeating_item_horiz" : "repeating_item";
    }
   
    JSONObject jsonObject = (JSONObject)json;
    try {
      Template t = cfg.getTemplate(template);
      StringWriter sw = new StringWriter(); 
      templateVars.put("itemNumber", itemNumber);
      templateVars.put("name", name);
      templateVars.put("rightItemText", rightItemText);
      templateVars.put("responseType", responseType);
      templateVars.put("unitLabel", unitLabel);
      templateVars.put("optionNames", appCodeLists.get(codeListOID));
      templateVars.put("multiSelectOptionNames", appMultiSelectLists.get(multiSelectListOID));
      templateVars.put("columns", columns);
      templateVars.put("responseLayout", responseLayout);
      templateVars.put("isInline", isInline);
      templateVars.put("mandatory", mandatory);
      templateVars.put("itemName", itemName);
      
      t.process(templateVars, sw);
      return sw.toString();
    } 
    catch (Exception e) {
      e.printStackTrace();
    } 
   return "";
  }

  public String getOID() { return OID; }
  public void setOID(String OID) { this.OID = OID; }

  public String getName() { return name; }
  public void setName(String name) { this.name = name; }

  public String getDataType() { return dataType; }
  public void setDataType(String dataType  ) { this.dataType = dataType; }
  
  public String getItemNumber() { return itemNumber; }
  public void setItemNumber(String itemNumber) { this.itemNumber = itemNumber; }
  
  public String getUnitLabel() { return unitLabel; }
  public void setUnitLabel(String unitLabel) { this.unitLabel = unitLabel; }
  
  public String getCodeListOID() { return codeListOID; }
  public void setCodeListOID(String codeListOID) { this.codeListOID = codeListOID; }
  
  public int getColumns() { return columns; }
  public void setColumns(int columns) { this.columns = columns; }
  
  public String getTemplate() { return template; }
  public void setTemplate(String template) { this.template = template; }
  
  public JSONObject getItemDetails() { return itemDetails; }
  public void setItemDetails(JSONObject itemDetails) { this.itemDetails = itemDetails; }

  public boolean isMandatory() { return mandatory; }
  public void setMandatory(boolean mandatory) { this.mandatory = mandatory; }

  public String getRightItemText() { return rightItemText; }
  public void setRightItemText(String rightItemText) { this.rightItemText = rightItemText; }

  public String getResponseType() { return responseType; }
  public void setResponseType(String responseType) { this.responseType = responseType; }

  public String getResponseLayout() { return responseLayout; }
  public void setResponseLayout(String responseLayout) { this.responseLayout = responseLayout; }

  public String isInline() { return isInline; }
  public void setInline(String isInline) { this.isInline = isInline; }

  public String getItemName() { return itemName; }
  public void setItemName(String itemName) { this.itemName = itemName; }

  public String getMultiSelectListOID() { return multiSelectListOID; }
  public void setMultiSelectListOID(String multiSelectListOID) { this.multiSelectListOID = multiSelectListOID; }
    
}  